zebra-state/src/in_memory/chains.rs

```diff
@@ -3,18 +3,98 @@ use crate::{Request, Response};
 use futures::prelude::*;
 use std::{
     pin::Pin,
+    sync::Arc,
     task::{Context, Poll},
 };
-use tower::Service;
+use tower::{Service, ServiceExt};
+use zebra_chain::{
+    block::{Block, BlockHeaderHash},
+    types::BlockHeight,
+};
+
+/// A representation of the chain state at a given height
+#[derive(Clone, Debug)]
+struct ChainState {
+    block: Arc<Block>,
+    // TODO: add more state here
```

Comment:
Some state we will probably want to store:

```suggestion
    // TODO: add more state here
    //  - SetMainTip: the cumulative proof of work
    //  - block.header.time: an im vector of previous block times
    //  - block.header.bits: an im vector of previous block difficulties
    //  - final_sapling_root_hash: hash bytes (Sapling/Blossom only)
    //  - light_client_hash: an im vector of previous subtree roots (Heartwood only)
    //  - state required to validate transactions
    
```

Reply:
I'd prefer to collapse these bits of information into the struct and then use `im` to track a vector of the structs, where only the block time for that block will be stored in each "ChainState" but the im vector of chain states will give you access to all the recent block times, difficulties, etc.

Reply:
I think you're right about `time` and `bits`. I don't want to duplicate data in recent blocks and in the context.

I think the `light_client_hash` set is different, because it could come from any Heartwood block (including those on disk). But the actual number of roots is very small. So I think the most efficient structure is an im vector of roots.

We should persist a single set of roots to disk, every time we persist a block. Then we can initialise the root set from disk when we load the verifier.

For more details, see the Heartwood section of:
https://github.com/ZcashFoundation/zebra/pull/742#discussion_r460380882

Reply:
Similarly, I think the most efficient way to work with the large UTXO and nullifier sets is an `im` map of deltas.

If we don't use `im` data structures, we could end up using a lot of memory by duplicating parts of these sets in each context.

Reply:
:thumbsup: